docs/metrics-api-design.md

```diff
@@ -0,0 +1,519 @@
+  * description, a brief, readable description of the metric instrument.
+  * unit, the unit of metric values following the UCUM convention
+  *       (https://unitsofmeasure.org/ucum.html).
+  * enabled, a boolean that turns on or off collection.
+  *
+  */ 
+  virtual nostd::unique_ptr<DoubleCounter> 
+                            NewDoubleCounter(nostd::string_view name, 
+                            nostd::string_view description,
+                            nostd::string_view unit, 
+                            nostd::string_view enabled);
+  
+ /*
+  * New int Counter
+  *
+  * Function that creates and returns a Counter metric instrument with value
+  * type int.
+  *
+  * Arguments:
+  * name, the name of the metric instrument (must conform to the above syntax).
+  * description, a brief, readable description of the metric instrument.
+  * unit, the unit of metric values following the UCUM convention
+  *       (https://unitsofmeasure.org/ucum.html).
+  * enabled, a boolean that turns on or off collection.
+  *
+  */ 
+  virtual nostd::unique_ptr<LongCounter> NewIntCounter(nostd::string_view name,
+                                                       nostd::string_view description,
+                                                       nostd::string_view unit,
+                                                       nostd::string_view enabled);
+  
+////////////////////////////////////////////////////////////////////////////////////
+//                                                                                //
+//                     Repeat above two functions for all                         //
+//                     six (five other) metric instruments.                       //
+//                                                                                //
+////////////////////////////////////////////////////////////////////////////////////
+  
+private:
+
+ /*
+  * CollectAll
+  *
+  * Checkpoints all metric instruments created from this meter and returns a
+  * vector of records containing the name, labels, and values of each instrument.
+  *
+  */
+  virtual nostd::vector<Record> CollectAll() = 0;
+  
+ /*
+  * RecordBatch
+  *
+  * Allows the functionality of acting upon multiple metrics with the same set 
+  * of labels with a single API call. Implementations should find bound metric 
+  * instruments that match the key-value pairs in the labels.
+  *
+  * Arugments:
+  * labels, labels associated with all measurements in the batch.
+  * values, a KeyValueIterable where the key is a string containing the name
+  *         of metric instruments such as "IntCounter" or "DoubleUpDownSumObserver"
+  *         and the value is the value to be recorded to all metric instruments in 
+  *         the batch of the associated instrument type.
+  *
+  */ 
+  virtual void RecordBatch(nostd::string_view labels, 
+                           nostd::KeyValueIterable values) noexcept;
```

Comment:
`KeyValueIterable` allows doubles, ints, strings, bools and list of those types as values. I think we should be more restrictive (as we only support doubles and ints for values here).

Reply:
That is an interesting point. What type of data structure would you suggest, given that we're restricted from using STL types?

Reply:
This has been changed to nostd::span<std::pair<nostd::shared_ptr<SynchronousInstrument, nostd::variant<int, double>>>.